### PR TITLE
Update TF fine-tuning docs

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR updates the fine-tuning sidebar tutorial with modern TF methods that were added in the most recent release.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
@stevhliu is there a better way to format a link to the `prepare_tf_dataset` docs [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18654/en/training#loading-data-as-a-tfdatadataset) than the way I did it here? It prints the whole `TFPreTrainedModel.prepare_tf_dataset` text, which looks a bit ugly on the page!

**Comment by stevhliu** [comment]
Yes, I believe you can just squiggly it!

[`~TFPreTrainedModel.prepare_tf_dataset`]

File: docs/source/en/training.mdx [hunk 1]
```diff
@@ -151,66 +151,113 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
...
->>> from transformers import TFAutoModelForSequenceClassification
-
->>> model = TFAutoModelForSequenceClassification.from_pretrained("bert-base-cased", num_labels=5)
+>>> tf_dataset = model.prepare_tf_dataset(dataset, batch_size=16, shuffle=True, tokenizer=tokenizer)
 ```
 
-Then compile and fine-tune your model with [`fit`](https://keras.io/api/models/model_training_apis/) as you would with any other Keras model:
+<Tip>
+
+The reason we pass the tokenizer to `prepare_tf_dataset` is so that it can correctly pad batches as they're loaded.
```
Comment (stevhliu) [inline docs/source/en/training.mdx:243]: This is a pretty substantial paragraph, so I think it's better to include it in the main body of content instead of as a Tip.

```suggestion
You need to pass the tokenizer to `prepare_tf_dataset` so it can correctly pad batches as they're loaded.
```

File: docs/source/en/training.mdx [hunk 2]
```diff
@@ -151,66 +151,113 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
...
+If all the samples in your dataset are the same length and no padding is necessary, you can skip this argument.
+If you need to do something more complex than just padding samples (e.g. corrupting tokens for masked language
+modelling), you can use the `collate_fn` argument instead to pass a function that will be called to transform the
+list of samples into a batch and apply any preprocessing you want. See our
+[examples](https://github.com/huggingface/transformers/tree/main/examples) or
+[notebooks](https://huggingface.co/docs/transformers/notebooks) to see this approach in action.
+
+</Tip>
+
+Once we've created a `tf.data.Dataset` we can simply compile and fit our model as before:
```
Comment (stevhliu) [inline docs/source/en/training.mdx:253]:
```suggestion
Once you've created a `tf.data.Dataset`, you can compile and fit the model as before:
```

**Review by stevhliu — COMMENTED** [review]
This is super cool, things are just getting easier and easier in TF! 🔥

Should we also update the finetune guides (like [this](https://huggingface.co/docs/transformers/v4.21.1/en/tasks/sequence_classification#train)) to use `prepare_tf_dataset`?

**Comment by Rocketknight1** [comment]
@stevhliu Thank you for the suggestions! I'll make more edits to incorporate the other bits and ping you again for a final look.

File: docs/source/en/training.mdx [hunk 3]
```diff
@@ -63,7 +63,7 @@ If you like, you can create a smaller subset of the full dataset to fine-tune on
 
 <a id='trainer'></a>
 
-## Train
+## Train with PyTorch Trainer
```
Comment (sgugger) [inline docs/source/en/training.mdx:66]: The Trainer is not from PyTorch, so this should be more like
```suggestion
## Train your PyTorch model with the Trainer
```

File: docs/source/en/training.mdx [hunk 4]
```diff
@@ -151,66 +151,112 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
...
+When you want to train a 🤗 Transformers model with the Keras API, you need to convert your dataset to a format that
+Keras understands. If your dataset is small, you can just convert the whole thing to NumPy arrays and pass it to Keras.
+
+First, load a dataset:
+
+```py
+from datasets import load_dataset
+
+dataset = load_dataset("glue", "cola")  # Simple text classification dataset
+dataset = dataset["train"]  # Just take the training split for now
```
Comment (sgugger) [inline docs/source/en/training.mdx:169]: Use the prose to explain what you're doing instead of comments in the code.

File: docs/source/en/training.mdx [hunk 5]
```diff
@@ -151,66 +151,112 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
...
+Next, load a tokenizer and tokenize the data as NumPy arrays:
+
+```py
+from transformers import AutoTokenizer
 
-The [`DefaultDataCollator`] assembles tensors into a batch for the model to train on. Make sure you specify `return_tensors` to return TensorFlow tensors:
+tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
+tokenized_data = tokenizer(dataset["text"], return_tensors="np", padding=True)
+
+labels = np.array(dataset["label"])  # Label is already an array of 0 and 1
```
Comment (sgugger) [inline docs/source/en/training.mdx:180]: Same here.

**Review by sgugger — APPROVED** [review]
Nice update!

File: docs/source/en/training.mdx (hunk 4 above)
Comment (Rocketknight1) [reply docs/source/en/training.mdx:169]: Done!

File: docs/source/en/training.mdx (hunk 5 above)
Comment (Rocketknight1) [reply docs/source/en/training.mdx:180]: Done!

File: docs/source/en/training.mdx (hunk 1 above)
Comment (Rocketknight1) [reply docs/source/en/training.mdx]: Done! I also included those edits.

File: docs/source/en/training.mdx [hunk 6]
```diff
@@ -151,66 +151,113 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
...
-Next, convert the tokenized datasets to TensorFlow datasets with the [`~datasets.Dataset.to_tf_dataset`] method. Specify your inputs in `columns`, and your label in `label_cols`:
+This approach is great when it works, but for larger datasets you might find it starting to become a problem. Why?
+Because the tokenized array and labels would have to be fully loaded into memory, and because Numpy doesn’t handle
+“jagged” arrays, so every tokenized sample would have to be padded to the length of the longest sample in the whole
+dataset. That’s going to make your array even bigger, and all those padding tokens will slow down training too!
+
+### Loading data as a tf.data.Dataset
+
+If you want to avoid this, you can load your data as a `tf.data.Dataset` instead. Although you can write your own
+`tf.data` pipeline if you want, we have two convenience methods for doing this - [`~datasets.Dataset.to_tf_dataset`]
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply docs/source/en/training.mdx]: Done!

**Comment by Rocketknight1** [comment]
@sgugger @stevhliu I finished incorporating your edits and did some other cleanup. I also replaced `to_tf_dataset` in the other fine-tuning pages. I didn't touch the translations, though - should I edit those too?

File: docs/source/en/training.mdx [hunk 7]
```diff
@@ -63,7 +63,7 @@ If you like, you can create a smaller subset of the full dataset to fine-tune on
 
 <a id='trainer'></a>
 
-## Train
+## Train your PyTorch model with the Trainer
```
Comment (sgugger) [inline docs/source/en/training.mdx:66]: No this section groups both training with PyTorch and TensorFlow with framework-specific blocks, so leave this as it was.

File: docs/source/en/training.mdx [hunk 8]
```diff
@@ -151,66 +151,113 @@ Then fine-tune your model by calling [`~transformers.Trainer.train`]:
 
 <Youtube id="rnTGBy2ax1c"/>
 
-🤗 Transformers models also supports training in TensorFlow with the Keras API.
+## Train with Keras
```
Comment (sgugger) [inline docs/source/en/training.mdx:154]: Same for this line.

**Review by sgugger — APPROVED** [review]
Thanks for iterating on this! No need to touch the translations, someone in the community might decide to update those later :-)

**Comment by Rocketknight1** [comment]
@sgugger I tried those edits but it looked a little odd because there was no separate header for the PyTorch section. I added a `Train` header to the whole thing with a brief intro, and then a `Train with PyTorch Trainer` header inside that block, which I think works a little better and makes it easier for people to find what they want in the sidebar. Let me know what you think!

**Comment by sgugger** [comment]
I'm not sure the headings will completely work with the framework content (for some reason those blocks don't appear in the PR previews so hard to know).

File: docs/source/en/training.mdx [hunk 9]
```diff
@@ -63,12 +63,18 @@ If you like, you can create a smaller subset of the full dataset to fine-tune on
 
 <a id='trainer'></a>
 
-## Train
-
 <frameworkcontent>
 <pt>
 <Youtube id="nvBXf7s7vTI"/>
 
+## Train
```
Comment (sgugger) [inline docs/source/en/training.mdx:70]: Train needs to be outside of the framework-content block.

File: docs/source/en/training.mdx [hunk 10]
```diff
@@ -63,12 +63,18 @@ If you like, you can create a smaller subset of the full dataset to fine-tune on
...
+
+At this point, you should follow the section corresponding to the framework you want to use. You can use the links
+in the right sidebar to jump to the one you want - and if you want to hide all of the content for a given framework,
+just use the button at the top-right of that framework's block!
+
```
Comment (sgugger) [inline docs/source/en/training.mdx:75]: Same for this intro.

**Review by stevhliu — APPROVED** [review]
Thanks for also updating all the finetune task guides with `prepare_tf_dataset`, this looks great! 👍

File: docs/source/en/training.mdx (hunk 10 above)
Comment (stevhliu) [reply docs/source/en/training.mdx:71]: Love this little intro for guiding user navigation/improving readability!

**Comment by Rocketknight1** [comment]
@sgugger :facepalm: I knew I'd regret pinging you before waiting for the job to finish and checking it myself. Fixed!

**Review by sgugger — APPROVED** [review]
Looks perfect now!
